### PR TITLE
Optionally record activities in the service

### PR DESCRIPTION
**Pull request by safesparrow**
Follow up on @baronfel 's initial work.

This PR adds activity recording to the FCS service.

It adds a dependency on `System.Diagnostics.DiagnosticSource` but nothing else.

Recording and exporting of activities can be enabled in user code (eg. in the IDE or external tests) - [example](https://github.com/safesparrow/fsharp-benchmark-generator/blob/otel_hacks/FCSBenchmark.Runner/Runner.fs#L65-L86).

## Sample traces
Sample traces produced can be found in https://github.com/safesparrow/fsharp-benchmark-generator/tree/main/trace_samples
They can be opened with Jaeger UI.

Jaeger UI screenshots:
<details>
<summary>Fantomas sample (<a href="https://github.com/fsprojects/fantomas">codebase</a>)</summary>

image
</details>

<details>
<summary>stress_huge_dense sample (<a href="https://github.com/dotnet/fsharp/tree/main/tests/projects/stress/huge_netstandard">codebase</a>)</summary>
 
 
image

</details>

## Sample traces for perf analysis of https://github.com/dotnet/fsharp/pull/13521
The driver for this PR is the need to easily verify the behaviour of the compiler/service, especially w.r.t. multi-threaded work.

With that in mind, I produced the following traces showcasing the behaviour of [parallel project analysis](https://github.com/dotnet/fsharp/pull/13521):
<details>
<summary>50_leaves sample, parallel analysis on, server GC</summary>

image
</details>
<details>
<summary>50_leaves sample, parallel analysis on, workstation GC</summary>

image
</details>

<details>
<summary>50_leaves sample, parallel analysis off, server GC</summary>

image
</details>

<details>
<summary>50_leaves sample, parallel analysis off, workstation GC</summary>

image
</details>

## Generating traces
Sample traces can be generated with the following tools:
- [Jaeger](https://www.jaegertracing.io/docs/1.37/getting-started/) as tracing backend and UI, and
- https://github.com/safesparrow/fsharp-benchmark-generator/ as a way to run benchmarks that produce traces
using the following script:
```
# Run Jaeger docker image - command from https://www.jaegertracing.io/docs/1.37/getting-started/
docker run -d --name jaeger -e COLLECTOR_ZIPKIN_HOST_PORT=:9411 -e COLLECTOR_OTLP_ENABLED=true   -p 6831:6831/udp   -p 6832:6832/udp   -p 5778:5778   -p 16686:16686   -p 4317:4317   -p 4318:4318   -p 14250:14250   -p 14268:14268   -p 14269:14269   -p 9411:9411   jaegertracing/all-in-one:1.37

# Install 'fcs-benchmark' global dotnet tool - https://github.com/safesparrow/fsharp-benchmark-generator
dotnet tool install fcsbenchmark --global --prerelease

# Run a benchmark of a sample solution, using local F# codebase, and 4 different GC&threading configurations
fcs-benchmark --sample 50_leaves --local=C:/projekty/fsharp/fsharp-otel -c C:/.artifacts --gc=Compare --parallel-analysis=Compare --record-otel-jaeger -w 0 -n 1
```

## Notes
Names of the activities and tags are likely wrong - please do suggest better names🙂 

To do:
- [x] Update surface baseline
- [x] Fix formatting
- [x] Attach traces for parallel vs sequential project analysis (https://github.com/dotnet/fsharp/pull/13521 )
- [ ] Verify activity and tag names
- [ ] More

### CODE REVIEW

File: NuGet.config
Comment (baronfel) [inline NuGet.config:16]: This is going to block merge, any new dependencies needed for this should be listed s that we can ask folks to add them to the internal feeds (which are public but just 'blessed' versions of packages to prevent version-substitution attacks from NuGet.org)

File: src/Compiler/Service/IncrementalBuild.fs
Comment (baronfel) [inline src/Compiler/Service/IncrementalBuild.fs:1371]: Is this rearrangement necessary? just looking for minimal diffs.

File: src/Compiler/Service/service.fs
Comment (baronfel) [inline src/Compiler/Service/service.fs:1530]: interesting - you had to move to explicit async CE usage to ensure correct scoping of the resource lifetimes?

File: src/fsc/fsc.targets
Comment (baronfel) [inline src/fsc/fsc.targets:55]: I think the most defensible addition here is the bare protocol - if end users want to consume this then they'll have to set up and configure the [collector](https://opentelemetry.io/docs/collector/). That keeps us endpoint-agnostic and only binding to the exporter protocol.

**Review by baronfel — COMMENTED** [review]
Nice - I think it might demonstrate the value better if you could attach trace views to show the results.

File: NuGet.config
Comment (safesparrow) [inline NuGet.config:16]: Need to delete this. Not sure if the `SystemDiagnosticsDiagnosticSourceVersion` package is available in other sources.

File: src/Compiler/Service/IncrementalBuild.fs
Comment (safesparrow) [reply src/Compiler/Service/IncrementalBuild.fs]: Sorry, this is leftover - at some point I needed `activity.Dispose()` before I extended the comp expression to support it implicitly. Will revert this line 👍

File: src/Compiler/Service/service.fs
Comment (safesparrow) [reply src/Compiler/Service/service.fs]: Same as above - needed it before I added support for `.Using()`.

File: src/fsc/fsc.targets
Comment (safesparrow) [reply src/fsc/fsc.targets]: Actually I'm going to remove this.
~This PR intends to only add activities in the service/background checker, not the compiler exe~.
This PR doesn't setup any exporters in the compiler exe the way your initial changes did - see `fscmain.fs`.

So the setup of the exporters happens in user code (eg. IDE, or external test code).

Have reverted these two lines as well.

Comment (safesparrow) [reply src/fsc/fsc.targets]: Long-term I think it would be best to allow users to specify an exporter via "late bindings" - as nasty as that sounds, I think it's better than asking users to setup one more backend service to forward the data.

But I hope this PR doesn't require making that decision.

File: NuGet.config
Comment (safesparrow) [reply NuGet.config]: Have removed now - we don't need it as we don't reference OpenTelemetry stuff anymore.

**Comment by safesparrow** [comment]
Have attached Jaeger screenshots now and removed any dependencies except for `System.Diagnostics.DiagnosticSource`.

File: tests/FSharp.Compiler.Service.Tests/FSharp.CompilerService.SurfaceArea.netstandard.expected
Comment (vzarytovskii) [inline tests/FSharp.Compiler.Service.Tests/FSharp.CompilerService.SurfaceArea.netstandard.expected:2296]: Should the activity be internal instead? Or does it have to be in public surface?

Comment (safesparrow) [reply tests/FSharp.Compiler.Service.Tests/FSharp.CompilerService.SurfaceArea.netstandard.expected:2296]: Good point. It could be internal. Will fix later today.

**Comment by vzarytovskii** [comment]
This is great! Few question:
- I presume it does not add much to the compiler runtime itself.
- I'm not very familiar with diagnosticsource, will it be adding anything to the etw/when tracing with something like dottrace/dotmemory/viewing in prefview if it's setup on the caller side?

**Comment by safesparrow** [comment]
> I'm not very familiar with diagnosticsource, will it be adding anything to the etw/when tracing with something like dottrace/dotmemory/viewing in prefview if it's setup on the caller side?

I think it can be used with `dotnet trace` according to https://www.meziantou.net/getting-telemetry-data-from-inside-or-outside-a-dotnet-application.htm

File: tests/FSharp.Compiler.Service.Tests/FSharp.CompilerService.SurfaceArea.netstandard.expected
Comment (safesparrow) [reply tests/FSharp.Compiler.Service.Tests/FSharp.CompilerService.SurfaceArea.netstandard.expected]: Done now 👍

**Comment by T-Gro** [comment]
@safesparrow I would be more than happy to get this in, would you mind if I offer a hand with getting it pass checks and resolving conflicts?

It would be helpful for me to have this in order to investigate more complex VS<->FCS interactions as well!

**Comment by vzarytovskii** [comment]
> EDIT: Correction - atm FSC doesn't depend on otel and doesn't set up recording. This means that there is currently no way for `fsc.exe`to record activities. We could keep it that way (which makes the PR simpler) and address fsc logging API separately, or decide on how to set it up. I don't have an opinion, only that it would be nice not to block merging the FCS part of it if possible.

This is fine, those should be solved separately.
Additionally, we would like to **not bring** any additional 3rd party dependencies in for FCS/FSC/FSI, since it will mean a lot of additional work for TPS/SBOM/approvals, increase in size, etc.

**Comment by safesparrow** [comment]
I merged @T-Gro 's changes in (thanks!).

(As I understand maintainers should have access to commit to the branch directly, but last time that didn't work for some reason, despite the "Allow edits..." checkbox being enabled.)

File: src/Compiler/Driver/ParseAndCheckInputs.fs
Comment (safesparrow) [inline src/Compiler/Driver/ParseAndCheckInputs.fs:1202]: Should `inputName` and `fileName` be actually the same field names? The only reason they are different I think is because I looked at the method name and chose `input` here rather than `file`.

File: src/Compiler/Facilities/Logger.fs
Comment (safesparrow) [inline src/Compiler/Facilities/Logger.fs:12]: Is `fsc` a good name here?

File: src/Compiler/Facilities/Logger.fs
Comment (safesparrow) [inline src/Compiler/Facilities/Logger.fs:14]: Should probably use `camelCase` (although I can see the `Logger.fsi` uses `PascalCase`).

File: src/Compiler/Facilities/Logger.fs
Comment (safesparrow) [inline src/Compiler/Facilities/Logger.fs:15]: Some formatting issues to be fixed.

File: src/Compiler/Checking/CheckDeclarations.fs
Comment (0101) [inline src/Compiler/Checking/CheckDeclarations.fs:5181]: Could we construct the Activity names using `nameof` instead of strings? In case we ever decide to rename any of these.

File: src/Compiler/Service/FSharpCheckerResults.fs
Comment (safesparrow) [inline src/Compiler/Service/FSharpCheckerResults.fs:2334]: Would it be useful to include `userOpName`? 🤔

File: src/Compiler/Service/FSharpCheckerResults.fs
Comment (safesparrow) [inline src/Compiler/Service/FSharpCheckerResults.fs:2495]: Would `projectFileName` be a useful tag?

File: src/Compiler/Service/service.fs
Comment (safesparrow) [inline src/Compiler/Service/service.fs:278]: Similarly here and above - should we use the same tag name for all things that are a project name/path? Or should we use the names of parameters in a given function? I don't mind either way.

File: src/Compiler/Service/service.fs
Comment (safesparrow) [inline src/Compiler/Service/service.fs:408]: Is this formatting Fantomas' default? 🤔

File: src/Compiler/Service/service.fs
Comment (safesparrow) [inline src/Compiler/Service/service.fs:470]:
```suggestion
                Activity.Start "BackgroundCompiler.ParseFile" [| "fileName", fileName; "UserOpName", userOpName; "cache", cache.ToString() |]
```

File: src/Compiler/Service/service.fs
Comment (safesparrow) [inline src/Compiler/Service/service.fs:497]:
```suggestion
                Activity.Start "BackgroundCompiler.GetBackgroundParseResultsForFileInProject" [| "fileName", fileName; "UserOpName", userOpName |]
```

File: src/Compiler/Service/service.fs
Comment (safesparrow) [inline src/Compiler/Service/service.fs:526]: Same casing issue here and in a few other places. We should make it consistent.

File: src/Compiler/Facilities/Logger.fs
Comment (safesparrow) [inline src/Compiler/Facilities/Logger.fs]: The `ActivityFacade` was brought in to seamlessly support a case where there are no listeners.
`activitySource.StartActivity(name)` returns `null` if there are no listeners.

Removing this causes NREs without listeners.

If such a facade is not desirable, an alternative would be to always hook up a dummy listener, or in other way support a scenario where no activity is returned.

Comment (safesparrow) [reply src/Compiler/Facilities/Logger.fs]: CC @T-Gro - we need to bring back support for no listeners in some way.

File: src/Compiler/Checking/CheckDeclarations.fs
Comment (safesparrow) [reply src/Compiler/Checking/CheckDeclarations.fs:5181]: The problem with that is it requires modules & function names to use `rec`.

File: src/Compiler/Facilities/Logger.fs
Comment (safesparrow) [reply src/Compiler/Facilities/Logger.fs]: Fixed now.

File: src/Compiler/Checking/CheckDeclarations.fs
Comment (T-Gro) [reply src/Compiler/Checking/CheckDeclarations.fs]: I started with this approach of turning functions recursive and rejected it.
When I see a function (or module) with rec, it immediately raises my awareness because it typically highlights and increase level of complexity and also theoretical computational power.
And the ability to render tracing strings does not seem to me like a strong justification.

File: src/Compiler/Service/service.fs
Comment (T-Gro) [reply src/Compiler/Service/service.fs:408]: I applied a full Fantomas check on the file. It should be deciding inline/linebreak depending on the width.

File: src/Compiler/Driver/ParseAndCheckInputs.fs
Comment (T-Gro) [reply src/Compiler/Driver/ParseAndCheckInputs.fs]: Seeing it again, I think it is better (from the perspective of looking at provided traces) to use same names across callsites, even if the arguments may vary slightly.
So well known names like:

fileName
project

Should be always the same.

File: src/Compiler/Facilities/Activity.fs
Comment (T-Gro) [inline src/Compiler/Facilities/Activity.fs:27]: Please check the PR into your fork.
We can avoid allocating the dummyDisposable, but it is needed for the internal computation expression's .Using to be defensive.
Basically, the .Using was implemented by calling .Dispose() without checking for nullness.

Comment (safesparrow) [reply src/Compiler/Facilities/Activity.fs]: >Please check the PR into your fork.

I already did. The problem was that the old code was doing `.AddTag` on a possibly null activity (which I since changed).

You're right that the `dummyDisposable` is not needed (although it's a singleton, so only allocated once).
I modified the code to return `null`. I wasn't aware that `using` (in C#) and `use` (in F#) handle `null`.

File: src/Compiler/Service/service.fs
Comment (T-Gro) [reply src/Compiler/Service/service.fs]: For reading produced traces, I prefer having names as consistent (e.g. always "project") as possible.

File: src/Compiler/Facilities/Activity.fs
Comment (T-Gro) [reply src/Compiler/Facilities/Activity.fs]: All good now, I see that also the CE's .Using is in the defensive version now.

**Review by T-Gro — DISMISSED** [review]

**Comment by T-Gro** [comment]
@safesparrow  I will have a look at the build failures related to package version being empty. For some reason the variables do not make their way into the CI build, have to investigate.

**Comment by T-Gro** [comment]
> @safesparrow I will have a look at the build failures related to package version being empty. For some reason the variables do not make their way into the CI build, have to investigate.

I think this might be it:
https://github.com/dotnet/fsharp/pull/14075/commits/2b9210f9b6c85c6016b01b68779709475001cb34

**Comment by vzarytovskii** [comment]
@KevinRansom @dsyme I'd like your review before we get it in.

This pretty much adds a new dependency to FCS (System.Diagnostics.DiagnosticSource, @safesparrow  may know how many new assemblies and what are their sizes), but will help us a lot in collecting additional info from FCS, when we run in LSP-like scenarios and benchmarks.

By default, nothing is collected or sent anywhere. 
Also, no PII or machine-related info is collected (@safesparrow may confirm it probably).

We will eventually be also able to hook it in to VS telemetry to have more granular info.

**Comment by baronfel** [comment]
Regarding the new dependency, System.Diagnostics.DiagnosticsSource is part of the Microsoft.NETCore.App Shared Framework, so every application will have it.  I think the most interesting question re: dependencies is how this interacts with Fable. @alfonsogarciacaro should we ifdef this work behind a feature flag that Fable wouldn't provide? Or is there an analog in the various JS runtimes that would make sense for Fable to polyfill with?

For FSAC this presents some really interesting visualization opportunities. The Activity and ActivitySource concepts are spreading like wildfire in the .NET ecosystem, and many libraries already use them for observability. The runtime provides counters via this mechanism. The StreamJsonRpc library (which is used both in VS and FSAC) uses this mechanism as well, creating a span for each LSP call. What this means is that we could get an end-to-end trace of not just the protocol level communication, but also the compiler-level work being doing for that call.

**Comment by alfonsogarciacaro** [comment]
Actually, the person who is maintaining the version of the F# compiler which is compilable to JS is @ncave, so they can give a better answer. But looking at the code, it's nice that `System.Diagnostics` it's not being used directly but through the `Activity` module. This should make it easy to polyfill it, and leave the call sites untouched.

> Note that ncave actually maintains two FCS forks: one is the actual FCS with a few modifications to improve incremental compilation in Fable tool, and the other is a stripped-down FCS version with many `#if FABLE_COMPILER` spread through the code that is used in the REPL (and maybe in the future in other runtimes).

**Comment by T-Gro** [comment]
@safesparrow :
I fixed all the checks in https://github.com/dotnet/fsharp/pull/14075 .
Can you merge from it?

It needed a package update, fixing one msbuild varialbe, and adding the nuget package to Directory.props of Vs integration.

(alternatively we can rebrand the PRs, but I think it would be better to migrate the fixes into yours and delete mine afterwards)

**Comment by T-Gro** [comment]
> Actually, the person who is maintaining the version of the F# compiler which is compilable to JS is @ncave, so they can give a better answer. But looking at the code, it's nice that `System.Diagnostics` it's not being used directly but through the `Activity` module. This should make it easy to polyfill it, and leave the call sites untouched.
> 
> > Note that ncave actually maintains two FCS forks: one is the actual FCS with a few modifications to improve incremental compilation in Fable tool, and the other is a stripped-down FCS version with many `#if FABLE_COMPILER` spread through the code that is used in the REPL (and maybe in the future in other runtimes).

Agree, polyfilling the two Activity.start* functions with diagnostical/debug messages to console would be best.

**Review by vzarytovskii — CHANGES_REQUESTED** [review]
Folks, please, don't merge it until @KevinRansom and @dsyme review it

File: src/Compiler/Facilities/Activity.fs
Comment (dsyme) [inline src/Compiler/Facilities/Activity.fs:1]: Looks to me like this should be in Utilities.  Basically 

* Utilities = General Utilities
* Facilities = Compiler-specific Utilities 

I don't mind if we rename these

* Utilities --> GeneralUtilities
* Facilities --> CompilerUtilities

This looks like a General utility - it would have use outside the compiler and is conceptually and implementation-wise independent of it.

BTW what's the `open Internal.Utilities.Library` used here for?

File: src/Compiler/Facilities/BuildGraph.fsi
Comment (dsyme) [inline src/Compiler/Facilities/BuildGraph.fsi:51]: Comment please

**Review by dsyme — DISMISSED** [review]

**Comment by dsyme** [comment]
I left a couple of minor comments

File: src/Compiler/Checking/CheckDeclarations.fs
Comment (auduchinok) [inline src/Compiler/Checking/CheckDeclarations.fs:5185]: It may probably also make sense to record the project file name or some other id of it, as a single file can be analysed multiple times in a context of several target frameworks, and it could be useful to distinguish them.

File: src/Compiler/Checking/CheckDeclarations.fs
Comment (auduchinok) [reply src/Compiler/Checking/CheckDeclarations.fs]: @dsyme this is a bit off-topic, but what do you think about always allowing recursive name usages inside nameof arguments? It looks like a good possible improvement to the feature to me, as it's one of the best uses for it.

File: src/Compiler/Checking/CheckDeclarations.fs
Comment (safesparrow) [reply src/Compiler/Checking/CheckDeclarations.fs:5185]: Nodes in the trace have parents and children. One can always see which project task this file-checking task belonged to, see eg: 
image

This doesn't currently include TargetFramework information - that could be added on the project-level nodes, alongside project path/name.

Comment (auduchinok) [reply src/Compiler/Checking/CheckDeclarations.fs:5185]: Each 'project + target framework' considered a different FCS project, so it covers everything needed.

Comment (safesparrow) [reply src/Compiler/Checking/CheckDeclarations.fs:5185]: Except that we don't currently tag TargetFramework, so we'd have two project nodes with the same project paths and no TF information.

Comment (auduchinok) [reply src/Compiler/Checking/CheckDeclarations.fs:5185]: > EDIT: I'm wrong - we tag project output path, which includes TF, so you're right 🙂Everything is included, if a bit hidden.

This is what we do in Rider, others probably rely on projects having different project stamps and outputs. It probably makes sense to tag something of these, if such projects would conflate otherwise?

We add the tfm to the project name because a few years ago it seemed like the best option for us for some reason which I can't remember anymore 🙂 Probably it was some more expensive equality checking or cache clearing when projects had the same path. VS used some other approach too prior to introducing stamps IIRC.

File: src/Compiler/Facilities/BuildGraph.fsi
Comment (safesparrow) [reply src/Compiler/Facilities/BuildGraph.fsi:51]: Done - please correct me if it's inaccurate.

File: src/Compiler/Facilities/Activity.fs
Comment (safesparrow) [reply src/Compiler/Facilities/Activity.fs]: I've done the move.

Also removed the open as it was unused.

**Comment by safesparrow** [comment]
I addressed comments with minor changes.
@vzarytovskii @dsyme Is there anything left to get this merged?

**Review by T-Gro — APPROVED** [review]

**Comment by vzarytovskii** [comment]
We can merge after @KevinRansom looks at it

**Review by dsyme — APPROVED** [review]

**Review by KevinRansom — CHANGES_REQUESTED** [review]
Thanks for this, it looks good.  However, we will need to do some due diligence internally to ensure that this satisfies our telemetry and privacy policies.  We will add this to our November milestone to see if we can get it approved.

**Comment by safesparrow** [comment]
> Thanks for this, it looks good.  However, we will need to do some due diligence internally to ensure that this satisfies our telemetry and privacy policies.  We will add this to our November milestone to see if we can get it approved.
> 
> Planning issue here:  https://github.com/dotnet/fsharp/issues/14210

Thanks for looking.

OOI what makes this more sensitive than any other existing logging? Also it doesn't depend on open telemetry and is off by default. Is there some increased risk of user data leakage or something like that?

**Comment by safesparrow** [comment]
@vzarytovskii If this gets merged, how difficult will it be to add new activities for other areas of the compiler in future PRs? I'm looking into optimization and having some detailed tracing there would help (eg. a span for each file being optimized, similar to how parsing & type-checking work atm).

Will it require new approvals?

**Comment by vzarytovskii** [comment]
> @vzarytovskii Vlad Zarytovskii FTE If this gets merged, how difficult will it be to add new activities for other areas of the compiler in future PRs? I'm looking into optimization and having some detailed tracing there would help (eg. a span for each file being optimized, similar to how parsing & type-checking work atm).
> 
> Will it require new approvals from the policy PoV?

Yep, it got approved in principle. Since compiler doesn't work with PII or EUII per se, we can just add new metrics, as long as we don't opt-in them by default (and we don't for neither old or new telemetry).

**Review by vzarytovskii — APPROVED** [review]

**Comment by vzarytovskii** [comment]
@safesparrow It's ready to merge, let me know if you want any more changes here, or if we can merge it now.

**Comment by T-Gro** [comment]
That is great news, let's merge it in so that new ones can build on this module.

**Comment by safesparrow** [comment]
I'm happy to have it merged, and we can iterate easily in future.

**Comment by vzarytovskii** [comment]
This needs to be reverted: https://github.com/dotnet/fsharp/issues/14413

**Comment by vzarytovskii** [comment]
Revert has to also go directly to 17.5

**Comment by vzarytovskii** [comment]
Or, as Kevin pointed out, that might be solved by adding package to vsix.

**Pull request merged**
